refactor lockstep programming model (#4840)
ComputationalRadiationPhysics/picongpu: docs/source/prgpatterns/lockstep.rst

.. sectionauthor:: René Widera, Axel Huebl

The *lockstep programming model* structures code that is evaluated collectively and independently by workers (physical threads) within a alpaka block.

> Actual processing is described by one-dimensional compile time known index domains of which can even be changed within a kernel.

Reviewer: ```suggestion
Actual processing is described by one-dimensional index domains which are known compile time and can even be changed within a kernel.
```

    auto oldVIdx = forEachExample(

>         [&](lockstep::Idx const idx) -> int32_t

Reviewer: ```suggestion
    auto oldElemIdx = forEachExample(
```
Also to be changed in a few other places here

Reviewer: Oops I meant in line#136

    }

> Non-Collective Loop oder particles

Reviewer: ```suggestion
Non-Collective Loop over particles
```

     * other workers) access that manipulated shared memory section
     */
    worker.sync();

Practical Examples
------------------

If possible kernels should be written without assuming any lockstep domain size and number of alpaka blocks selected at the kernel start.

> This ensure that the kernel results are always correct even if the user not chose the right parameters for the kernel execution.

Reviewer: ```suggestion
This ensure that the kernel results are always correct even if the user doesn't chose the right parameters for the kernel execution.
```

  .. literalinclude:: ../../../include/pmacc/test/lockstep/lockstepUT.cpp
     :language: C++
     :start-after: doc-include-start: lockstep generic kernel
     :end-before: doc-include-end: lockstep generic kernel
     :dedent:

The block domain size can also be derived from a instance of any object if the trait ``pmacc::lockstep::traits::MakeBlockCfg`` is defined.

  .. literalinclude:: ../../../include/pmacc/test/lockstep/lockstepUT.cpp
     :language: C++
     :start-after: doc-include-start: lockstep generic kernel buffer selected domain size
     :end-before: doc-include-end: lockstep generic kernel buffer selected domain size
     :dedent:

> Sometimes it is not possible to write a generic the kernel and a hard coded block domain size is required to fulfill stencil condition or other requirements.

Reviewer: ```suggestion
Sometimes it is not possible to write a generic kernel and a hard coded block domain size is required to fulfill stencil condition or other requirements.
```

> In this case it is possible to use in on device ``pmacc::lockstep::makeForEach<hardCodedBlockDomSize>(wroker)``.

Reviewer: ```suggestion
In this case it is possible to use on device ``pmacc::lockstep::makeForEach<hardCodedBlockDomSize>(wroker)``.
```

The problem is that the user needs to know this hard coded requirement during the kernel call else it could be the kernel is running slow.

> To many worker threads are idling during the execution because the selected block domain during the kernel call is larger than the required block domain within the kernel.

Reviewer: ```suggestion
It is possible that too many worker threads are idling during the execution because the selected block domain during the kernel call is larger than the required block domain within the kernel.
```
